src/main/scheduler.c

```diff
@@ -0,0 +1,288 @@
+    },
+#endif
+
+#ifdef TELEMETRY
+    [TASK_TELEMETRY] = {
+        .taskName = "TELEMETRY",
+        .taskFunc = taskTelemetry,
+        .desiredPeriod = 1000000 / 250,         // 250 Hz
+        .staticPriority = TASK_PRIORITY_IDLE,
+    },
+#endif
+
+#ifdef LED_STRIP
+    [TASK_LEDSTRIP] = {
+        .taskName = "LEDSTRIP",
+        .taskFunc = taskLedStrip,
+        .desiredPeriod = 1000000 / 100,         // 100 Hz
+        .staticPriority = TASK_PRIORITY_IDLE,
+    },
+#endif
+};
+
+void getTaskInfo(cfTaskId_e taskId, cfTaskInfo_t * taskInfo)
+{
+    taskInfo->taskName = cfTasks[taskId].taskName;
+    taskInfo->desiredPeriod = cfTasks[taskId].desiredPeriod;
+    taskInfo->staticPriority = cfTasks[taskId].staticPriority;
+    taskInfo->maxExecutionTime = cfTasks[taskId].maxExecutionTime;
+    taskInfo->totalExecutionTime = cfTasks[taskId].totalExecutionTime;
+}
+
+void rescheduleTask(cfTaskId_e taskId, uint32_t newPeriodMicros)
+{
+    if (taskId == TASK_SELF)
+        taskId = currentTaskId;
+
+    if (taskId < TASK_COUNT) {
+        cfTasks[taskId].desiredPeriod = MAX(100, newPeriodMicros);  // Limit delay to 100us (10 kHz) to prevent scheduler clogging
+    }
+}
+
+void scheduler(void)
+{
+    uint8_t taskId;
+    uint8_t selectedTaskId;
+    uint8_t selectedTaskDynPrio;
+
+    static uint32_t timeCpuBusy = 0;
+    static uint32_t timeCpuIdle = 0;
+    static uint32_t previousSchedulerTime = 0;
+
+    /* Cache currentTime */
+    currentTime = micros();
+
+    /* Update task dynamic priorities */
+    for (taskId = 0; taskId < TASK_COUNT; taskId++) {
+        uint16_t newDynamicPriority = 0;
+        uint16_t taskAgeCycles = ((currentTime - cfTasks[taskId].lastExecutedAt) / cfTasks[taskId].desiredPeriod);
+
+        /* Task has checkFunc - event driven */
+        if (cfTasks[taskId].checkFunc != NULL) {
+            /* Increase priority for event driven tasks */
+            if (cfTasks[taskId].checkFunc()) {
+                // FIXME: need cleaner approach here
+                /* If task is already scheduled to execution by event, force time-based scheduling to avoid task starvation */
+                if (cfTasks[taskId].dynamicPriority > 0) {
+                    newDynamicPriority = 1 + cfTasks[taskId].staticPriority * (1 + taskAgeCycles);
+                }
+                else {
+                    newDynamicPriority = 1 + cfTasks[taskId].staticPriority;
+                }
+            }
+        }
+        /* Task is time-driven, dynamicPriority is last execution age measured in desiredPeriods) */
+        else {
+            if (taskAgeCycles) {
+                newDynamicPriority = 1 + cfTasks[taskId].staticPriority * taskAgeCycles;
+            }
+        }
+
+        /* limit new priority to avoid overflow of uint8_t */
+        cfTasks[taskId].dynamicPriority = MIN(newDynamicPriority, 255);
+    }
+
+    /* Find the task with highest dynamicPriority - would be executed next */
```

Comment:
Why not use an ordered list (by dynamic priority) to keep track of the task to be run? Here you would just take the head of the list and in the "update task dynamic priorities" loop above, you would sort the list appropriately. Not sure it will be better/faster though. Just my 2 cent.

P.S. I'm new to cleanflight (and even haven't flighted it yet) but this scheduler thing looks really interesting.